### PR TITLE
Support all kinds of errors (errors, exits, throws)

### PR DESCRIPTION
**Pull request by whatyouhide**
Heavily inspired by #18, tries to solve #14.

\cc @lexmag

### CODE REVIEW

**Comment by whatyouhide** [comment]
@lexmag updated with the stuff we discussed today.

File: lib/rollbax.ex [hunk 1]
```diff
@@ -28,10 +28,22 @@ defmodule Rollbax do
...
-  def report(exception, stacktrace, meta \\ %{} , occurr_data \\ %{})
-  when is_list(stacktrace) and is_map(meta) and is_map(occurr_data) do
-    message = Exception.format(:error, exception, stacktrace)
+  def report(kind, value, stacktrace, meta \\ %{}, occurr_data \\ %{})
+  when kind in [:error, :exit, :throw] and is_list(stacktrace) and is_map(meta) and is_map(occurr_data) do
+    # We need this manual check here otherwise Exception.format_banner(:error,
+    # term) will assume that term is an Erlang error (it will say
+    # "** # (ErlangError) ...").
+    if kind == :error and not Exception.exception?(value) do
+      raise ArgumentError, "excpected an exception when the kind is :error, got: #{value}"
```
Comment (philss) [inline lib/rollbax.ex]: there is a minor typo here:  `excpected`.

File: lib/rollbax.ex [hunk 2]
```diff
@@ -28,10 +28,22 @@ defmodule Rollbax do
...
+    end
+
+    body = Rollbax.Item.body_for_trace(kind, value, stacktrace)
     meta = Map.put(meta, :rollbax_occurr_data, occurr_data)
-    Rollbax.Client.emit(:error, message, meta)
+    Rollbax.Client.emit(:error, unix_timestamp(), body, meta)
+  end
+
+  defp unix_timestamp() do
```
Comment (lexmag) [inline lib/rollbax.ex]: Let's name it `unix_time`.

Comment (whatyouhide) [inline lib/rollbax.ex]: Updated 👍

File: test/rollbax/client_test.exs [hunk 3]
```diff
@@ -38,9 +38,14 @@ defmodule Rollbax.ClientTest do
...
     :ok = RollbarAPI.stop
     log = capture_log(fn ->
-      :ok = Client.emit(:error, "miss", %{})
+      :ok = Client.emit(:error, unix_timestamp(), %{"message" => %{"body" => "miss"}}, %{})
     end)
     assert log =~ "[error] (Rollbax) connection error:"
     refute_receive {:api_request, _body}
   end
+
+  defp unix_timestamp() do
```
Comment (lexmag) [inline test/rollbax/client_test.exs]: `time`

File: lib/rollbax/item.ex [hunk 4]
```diff
@@ -15,20 +15,52 @@ defmodule Rollbax.Item do
...
+    |> put_location(location)
+  end
+
+  def stacktrace_entry_to_frame({fun, arity, location}) do
+    %{"method" => Exception.format_fa(fun, arity)}
+    |> put_location(location)
+  end
+
+  defp put_location(frame, location) do
+    frame = if file = location[:file], do: Map.put(frame, "filename", List.to_string(file)), else: frame
```
Comment (lexmag) [inline lib/rollbax/item.ex]: I think we should put info only if file and line present together.

File: lib/rollbax/item.ex [hunk 5]
```diff
@@ -15,20 +15,71 @@ defmodule Rollbax.Item do
...
+  end
+
+  def stacktrace_entry_to_frame({module, fun, arity, location}) when is_list(arity) do
+    %{"method" => Exception.format_mfa(module, fun, length(arity)), "args" => Enum.map(arity, &inspect/1)}
+    |> put_location(location)
+  end
+
+  def stacktrace_entry_to_frame({fun, arity, location}) when is_integer(arity) do
+    %{"method" => Exception.format_fa(fun, arity)}
+    |> put_location(location)
```
Comment (lexmag) [inline lib/rollbax/item.ex]: Seems better to move `put_location` call into `stacktrace_to_frames`.

File: lib/rollbax/item.ex [hunk 6]
```diff
@@ -15,20 +15,71 @@ defmodule Rollbax.Item do
...
       |> put_custom(meta)
       |> Map.put("level", level)
-      |> Map.put("timestamp", time)
+      |> Map.put("timestamp", timestamp)
     end)
   end
 
-  defp put_body(data, msg) do
-    Map.put(data, "body", %{"message" => %{"body" => msg}})
+  def body_for_message(message) do
```
Comment (lexmag) [inline lib/rollbax/item.ex]: Maybe `message_to_body` and `exception_to_body` (that one reads much better for sure)?

File: lib/rollbax/notifier.ex [hunk 7]
```diff
@@ -44,4 +44,10 @@ defmodule Rollbax.Notifier do
     %{level: Keyword.get(config, :level, :error),
       metadata: Keyword.get(config, :metadata, [])}
   end
+
+  defp ts_to_unix({{_yy, _mm, _dd} = date, {h, m, s, _ms}}) do
+    now = :calendar.datetime_to_gregorian_seconds({date, {h, m, s}})
+    epoch = :calendar.datetime_to_gregorian_seconds({{1970, 1, 1}, {0, 0, 0}})
```
Comment (lexmag) [inline lib/rollbax/notifier.ex]: `@unix_epoch :calendar.datetime_to_gregorian_seconds {{1970, 1, 1}, {0, 0, 0}}`

File: lib/rollbax/notifier.ex [hunk 8]
```diff
@@ -44,4 +44,10 @@ defmodule Rollbax.Notifier do
     %{level: Keyword.get(config, :level, :error),
       metadata: Keyword.get(config, :metadata, [])}
   end
+
+  defp ts_to_unix({{_yy, _mm, _dd} = date, {h, m, s, _ms}}) do
```
Comment (lexmag) [inline lib/rollbax/notifier.ex]: We have four positions for a year value, let's go with `{{_, _, _} = date, {hour, min, sec, _}}`.

File: lib/rollbax/notifier.ex [hunk 9]
```diff
@@ -44,4 +44,10 @@ defmodule Rollbax.Notifier do
...
+    now = :calendar.datetime_to_gregorian_seconds({date, {h, m, s}})
```
Comment (lexmag) [inline lib/rollbax/notifier.ex]: It's not `now`, rather `event_time` (`log_time`?), but having `@unix_epoch` we probably don't even need to name it. :bowtie:

File: lib/rollbax/notifier.ex (hunk 8 above)
Comment (whatyouhide) [inline lib/rollbax/notifier.ex]: @lexmag I used `{hour, min, sec, _millisec}` here, I'm not a fan of just `_` for ms. Wdyt?

**Comment by whatyouhide** [comment]
@lexmag updated everything, only did one thing differently (and left a comment there).

File: lib/rollbax/item.ex [hunk 10]
```diff
@@ -15,20 +15,78 @@ defmodule Rollbax.Item do
...
     end)
   end
 
-  defp put_body(data, msg) do
-    Map.put(data, "body", %{"message" => %{"body" => msg}})
+  def message_to_body(message) do
+    %{"message" => %{"body" => message}}
+  end
+
+  def exception_to_trace(kind, value, stacktrace) do
```
Comment (lexmag) [inline lib/rollbax/item.ex]: `_to_body`?

File: lib/rollbax/notifier.ex (hunk 8 above)
Comment (lexmag) [inline lib/rollbax/notifier.ex]: 👍

File: lib/rollbax/item.ex (hunk 10 above)
Comment (whatyouhide) [inline lib/rollbax/item.ex]: Whoops, confused the two "body"ies. Updated.

**Pull request merged**